metal/helpers_device.go

```diff
@@ -110,25 +110,43 @@ func getPorts(ps []packngo.Port) []map[string]interface{} {
 	return ret
 }
 
-func waitUntilReservationProvisionable(id string, meta interface{}) error {
+const (
+	deprovisioning = "deprovisioning"
+	provisionable  = "provisionable"
+	reprovisioned  = "reprovisioned"
+	errstate       = "error"
+)
+
+func hwReservationStateRefreshFunc(client *packngo.Client, reservationId, instanceId string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		r, _, err := client.HardwareReservations.Get(reservationId, nil)
+		state := deprovisioning
+		switch {
+		case err != nil:
+			err = friendlyError(err)
+			state = errstate
+		case r != nil && r.Provisionable:
+			state = provisionable
+		case r != nil && r.Device != nil && (r.Device.ID != "" && r.Device.ID != instanceId):
+			state = reprovisioned
```

Comment:
We should log when a hardware reservation was reprovisioned to another instance while awaiting deprovision.  This is a good behavior that will allow Terraform to exit at the right time while waiting for deprovision, but it will be confusing to users that expected a full deprovision (likely awaiting a reprovision of their own).